# Stahl et al., 2024, Meta-analysis of genome-wide CRISPR screens reveals RAB9A as a SARS-CoV-2 dependency factor involved in viral entry

[![DOI](https://TBA)

_The paper this repository relates to is currently under preparation._

This directory contains scripts and files supporting the publication: </br>

[Stahl _et al._](http://LINK_TBA), 2024, **Meta-analysis of genome-wide CRISPR screens reveals RAB9A as a SARS-CoV-2 dependency factor involved in viral entry**, _Journal_TBA_

## Abstract

<div style="text-align: justify; vertical-align: middle;">
Despite recent advances in the development of vaccines and antivirals, COVID-19 remains a thread to public health. In order to reveal dependency factors promoting SARS-CoV-2 infection, we conducted two genome-wide CRISPR knock-out screens in the human lung epithelial cell lines A549 and Calu-3. Integrating our data with data from four published CRISPR screens performed in different cells identified 24 potential dependency factors, including the ACE2 receptor of SARS-CoV-2. The overlapping factors included RAB9A, a GTPase involved in retrograde trafficking from late endosomes, which was selected for subsequent validation. We confirmed the importance of RAB9A for SARS-CoV-2 infection by knock-down in Huh7 HeLa ACE2 and A549 ACE2 cells, cytotoxicity assays, and expression analysis in relevant tissues. Pseudovirus experiments revealed an involvement of RAB9A in entry that was mitigated upon ectopic expression of the host protease TMPRSS2, suggesting that RAB9A exclusively promotes the endocytic entry of SARS-CoV-2. By dissecting the different steps of SARS-CoV-2 entry, we excluded potential effects of RAB9A on virus binding and were able to show an increased rate of virus internalization followed by retention of virus in lysosomes of RAB9A-depleted cells. Fusion assays revealed that RAB9A supports the fusion of virions with the endosomal membrane. This mechanism seems to be supported by the RAB9A-specific guanine nucleotide exchange factor DENND2D, indicating that GTPase activity is required for its proviral function. 
</div>

## Contact

Should you encounter any issues or have any questions regarding code and data contained in this repository, please contact Yannick Stahl <yannick.stahl@med.uni-heidelberg.de>, Florian Heigwer <f.heigwer@dkfz.de>, Michael Boutros <m.boutros@dkfz.de> or Ralf Bartenschlager <ralf.bartenschlager@med.uni-heidelberg.de>.

Raw sequencing read data are deposit at ENA under the Study ID: PRJEB72772
